imbi/clients/aws.py

```diff
@@ -11,6 +11,18 @@ async def get_credentials(session: aioboto3.Session, role_arn: str,
             WebIdentityToken=id_token)
 
 
+async def get_parameter(name: str, session: aioboto3.Session,
```

Comment:
I wish that I could comment directly on the commit ... sigh. Anyway, what do you think about making an AWS client class instead of a bunch of module level functions that all take `session`, `access_key_id`, `session_token`? I was thinking something like the following. Probably not a _right now_ sorta thing but I think that it would be a useful refactor at the end of the day.

```python
from __future__ import annotations

import enum
import typing

import aioboto3


class Credentials(typing.TypedDict):
    access_key_id: str | None
    secret_access_key: str | None
    session_token: str | None


class ParameterType(str, enum.Enum):
    STRING: 'String'
    SECURE_STRING: 'SecureString'
    STRING_LIST: 'StringList'


class AWSClient:
    def __init__(self, *, session: aioboto3.Session | None = None) -> None:
        super().__init__()
        self.session = aioboto3.Session() if session is None else session
        self._auth = Credentials(access_key_id=None, secret_access_key=None,
                                 session_token=None)

    async def assume_role(self, role_arn, username, id_token) -> dict:
        async with self.session.client('sts') as client:
            rsp = await client.assume_role_with_web_identity(
                RoleArn=role_arn, RoleSessionName=f'{username}-imbi',
                WebIdentityToken=id_token
            )
            creds = rsp['Credentials']
            self._auth.update({
                'access_key_id': creds['AccessKeyId'],
                'secret_access_key': creds['SecretAccessKey'],
                'session_token': creds['SessionToken'],
            })

    def set_credentials(self, credentials: Credentials) -> None:
        self._auth.update(credentials)

    async def get_parameters_by_path(self, path: str) -> list:
        async with self.session.client('ssm', **self._auth) as client:
            params = []
            paginator = client.get_paginator('get_parameters_by_path')
            async for page in paginator.paginate(Path=path,
                                                 Recursive=True,
                                                 WithDecryption=True,
                                                 MaxResults=10):
                params.extend(page['Parameters'])
        return params

    async def put_parameter(self, path: str, param_type: ParameterType,
                            value: str, *, overwrite: bool = False) -> None:
        ...

    async def delete_parameter(self, path: str) -> None:
        ...
```

Reply:
It's a little more complicated because we cache the credentials for the user in Redis. I suppose we could inject Redis into the class as well.

Reply:
I tried giving this a go. First injecting Redis, but this seems shoehorned. Then I pulled out the `assume_role` to just return creds in a function like it does now, and then inject them into the class. That worked, but only made everything a bit more verbose. We only ever call one method on the object at a time before throwing it away. Something like this would be useful if we're performing multiple operations in the same environment at a time. I think I'd like to hold off on doing something like this till that time so we have a clearer vision of how it can constructed to be useful.